floats/floats.go

```diff
@@ -515,31 +515,109 @@ func NaNPayload(f float64) (payload uint64, ok bool) {
 // Nearest returns the index of the element in s
 // whose value is nearest to v.  If several such
 // elements exist, the lowest index is returned.
-// Panics if len(s) == 0.
+// Nearest panics if len(s) == 0.
 func Nearest(s []float64, v float64) int {
+	if len(s) == 0 {
+		panic("floats: zero length slice")
+	}
+	switch {
+	case math.IsNaN(v):
+		return 0
+	case math.IsInf(v, 1):
+		return MaxIdx(s)
```

Comment:
There is an issue with `MaxInd` as it does not handle NaNs "properly" in case first element is NaN(). Following tests will fail:
```
		{
			in:    []float64{math.NaN(), 0, -1, 2},
			query: math.Inf(1),
			want:  3,
			desc:  "Wrong index returned when query is +Inf and no +Inf element exists",
		},
		{
			in:    []float64{math.NaN(), 0, -1, 2},
			query: math.Inf(-1),
			want:  2,
			desc:  "Wrong index returned when query is -Inf and no -Inf element exists",
		}
```

Changing [this](https://github.com/gonum/gonum/blob/master/floats/floats.go#L436) to something like this
```
	var (
		max float64
		idx int
	)
	allnan := true
	for k, v := range x {
		if !math.IsNaN(v) {
			allnan = false
			idx, max = k, v
			break
		}
	}
	if allnan {
		return 0
	}
```

Reply:
Fixed.